### ISSUE TITLE
upgrade syncthing option

### ISSUE THREAD

**Issue by buttle** [syncthing/syncthing-gtk#290]
Syncthing would be easier to use for new users if syncthing-gtk upgraded syncthing to new versions.
We have been commenting on it here.

https://forum.syncthing.net/t/with-me-syncthing-gtk-is-still-with-syncthing-version-13-6/7712

For many people it's hard enough to understand IDs, p2p, and other concepts. If we can save them the extra difficulty of learning how to upgrade that would be a big help.
Maybe a 'Check for upgrade' button?

Cheers!

**Comment by kozec** [comment]
You mean something like this?

![img](https://cloud.githubusercontent.com/assets/1282500/16953954/5212df66-4dcf-11e6-8c73-61d459d56d3c.png)

**Comment by buttle** [comment]
Hi,
Yes, I've got that checked. And it upgrades. (It checks every 12 hours right?)
But when there is a new major version, I haven't seen any upgrade happening. That's normal behaviour, isn't it?
So, If I'm using syncthing-gtk I need to stop it, run syncthing -upgrade, and then start syncthing-gtk again. Is that the procedure? A 'Manual upgrade' button would be good for avoiding that and making syncthing easier for inexperienced users.

Chris.

**Comment by kozec** [comment]
It doesn't really work like that. ST-GTK will upgrade Syncthing to any version it can handle. When there is new Syncthing release introducing API changes, I'll release new ST-GTK version and only that new version will update Syncthing. It would most likely crash otherwise.

Currently ST-GTK v0.9.0.0 - v0.9.0.3 will download Syncthing v0.13.10, v0.8 - v0.8.3 will download Syncthing v0.12.something and for Syncthing v0.14.0, you need ST-GTK v0.9.1.

**Comment by Coeur-Noir** [comment]
I'm running syncthing-gtk 0.9.0.3 here but syncthing daemon is 0.14 and I have [some issues](url=https://forum.syncthing.net/t/v0-14-brings-a-surprise-previously-deleted-files-come-re-appear/7814) ( may be not relative to syncthing-gtk ).
Where can I find syncthing-gtk 0.9.1 ?
Running ubuntu 14.04 on 3 machines, webupd8 ppa enabled…

**Comment by kozec** [comment]
> I'm running syncthing-gtk 0.9.0.3 here but syncthing daemon is 0.14 and I have some issues

That doesn't sounds like something ST-GTK could possibly do :)

**Comment by buttle** [comment]
Thanks @kozec 

How do ST-GTK updates get pushed out. There are different install options for ST_GTK so I'm wondering.. Will a windows user receive notification of a ST_GTK update (installer and/or chocolate) the same times as let's say, a Suse or Ubuntu user? What about installations done with the .deb or tarballs?
What are the scenarios where a user will not be notified of a new ST_GTK version?

Chris.

**Comment by kozec** [comment]
> What are the scenarios where a user will not be notified of a new ST_GTK version?

Basically, if he is using Windows or anything else without package manager. ST-GTK doesn't checks for it's own updates.

**Comment by buttle** [comment]
Would it be possible to programmer ST-GTK to check for a ST-GTK update every now and then? 

I'm just imagining a larger group of users without much tech knowledge. When a new version of syncthing comes out, the upgrade of the whole group needs to be easy or else things will stop working.

Chris.

**Comment by kozec** [comment]
> Would it be possible to programme ST-GTK to check for a ST-GTK update every now and then?

Dunno, wouldn't that just replicate work of package manager? Plus, not knowing which distro is user using, it could display only something like "There is new version available, refresh your package manager and hope it's already there" :)

**Comment by buttle** [comment]
Not everyone uses a package manager. Package managers can also get 'stuck' on some dependency and not work until some IT person fixes it.

I can imagine this happening.
1. Syncthing releases a new major version
2. Web interface users upgrade and become incompatible with other peers
3. Some users will get a package manager notification. (could be tomorrow, could be next month).
4. Other users never get a notification.
5. Group becomes broken.

ST-GTK to the rescue.
1. st-gtk notifies of st-gtk update.
2. user updates st-gtk
3. st-gtk notifies of syncthing update
4. user updates syncthing
5. everybody happy.

Don't know what you think about that. :)
Chris.

**Comment by Coeur-Noir** [comment]
I'm exactly in the situation, where syncthing daemon has been upgraded to 0.14 before syncthing-gtk 0.9.1 was available in usual repo…

At least, if st-gtk 9.0.3 could not handle st-daemon 0.14 why then did it « run » ? It should have stopped itself with a warning « not compatible or not tested with daemon » ?
![capture](https://cloud.githubusercontent.com/assets/1895947/16984586/36fe8444-4e7b-11e6-8f40-8e411bfe140b.png)
![crash](https://cloud.githubusercontent.com/assets/1895947/16984590/3d946df0-4e7b-11e6-8f04-4fab318e0235.png)

**Comment by kozec** [comment]
@buttle 

Actually, in "Package managers can also get 'stuck' on some dependency and not work until some IT person fixes it." situation, letting ST-GTK to autoupdate itself or letting user to update it manually would most likely break it completely :)

@Coeur-Noir 
> At least, if st-gtk 9.0.3 could not handle st-daemon 0.14 why then did it « run » 

For two reasons:
- st-daemon version becomes known only after it's started, so ST-GTK can't know which version is it going to start
- it's kinda hard to store last supported version beforehand in ST-GTK release without time machine :)

**Comment by buttle** [comment]
>  letting ST-GTK to autoupdate itself or letting user to update it manually would most likely break it completely :)

Yes, I'm aware of that. :)
So maybe package managers aren't the best way to distribute ST-GTK?
Maybe a ST-GTK wrapper app that gets installed by the package manager?
I really don't know what would be the best approach. But the problem exists.

Chris.

**Comment by Coeur-Noir** [comment]
I had that same crash randomly on 2 out of the 3 machines since updating to st-gtk 0.9.1

I think there is a real problem here : 
![capture du 2016-07-20 14 28 09](https://cloud.githubusercontent.com/assets/1895947/16986464/e6d291ca-4e86-11e6-884e-9b0ec11808d6.png)

**Comment by kozec** [comment]
Ok, I don't think we are going anywhere, so how about this:

If ST-GTK finds there is new Syncthing version that it cannot handle without update, it could display message like this:

![img](https://cloud.githubusercontent.com/assets/1282500/17014056/a502b684-4f22-11e6-9aab-6e5dd7aad4cc.png)


For stuff installed from OBS, it can even display download link to latest release, but I really don't like that idea.

**Comment by buttle** [comment]
What is OBS?
Looks good, A big improvement!
I still don't think we can expect users to manually reinstall (every month or so!) an application that they have already installed. People are lazy, or they're not techno-freaks, or they don't know how, or repetitive installation doesn't inspire confidence, or for whatever other reason.
I feel ST-GKT needs to maintain folders synchronised. I don't really know how to help you out. :(

**Comment by kozec** [comment]
> What is OBS?

[OpenSUSE Build Service](http://software.opensuse.org/download.html?project=home%3Akozec&package=syncthing-gtk), thing that builds ST-GTK packages and repositories for rpm-based distros. 

> I still don't think we can expect users to manually reinstall

This may be a stupid question, but how did you installed ST-GTK for first time? I'm not aware of any distro/package manager that couldn't do updates automatically and AFAIK, most of them even doing automatic updates by default.

**Comment by Coeur-Noir** [comment]
I should have started another discussion → enable "ignore permission bits" on each sync'd folder has immediately solved the problems. I'm not sure why though.
___________

> how did you installed ST-GTK for first time?
Using Ubuntu here I may not be the best example. I've added the 2 repositories webupd8 ppa & apt.syncthing… so both are automatically updated.

**Comment by buttle** [comment]
> how did you installed ST-GTK for first time? 
I don't remember how I did it on my laptop. Recently was at an office space where one person had already installed it on one linux PC via la deb file, I installed it on their second PC (ubuntu) with via the repository, and we installed it on the windows machine they have with the windows installer. That office space will soon have a dysfunctional syncthing installation. That's sort of my point here.

**Comment by buttle** [comment]
@kozec Hi.
Your reason for not upgrading major versions of ST is because that might make changes to the REST API and that would break ST-GTK, is that right?

If that is correct then how about this?
1. A menu item somewhere to 'Force syncthing upgrade to latest version'. That means everyone can be sure they are on the same version of syncthing by just clicking a menu item.
2. If / when a new ST version breaks the API, then include
https://cloud.githubusercontent.com/assets/1282500/17014056/a502b684-4f22-11e6-9aab-6e5dd7aad4cc.png

How does that seem to you?
Chris.

**Comment by kozec** [comment]
> That office space will soon have a dysfunctional syncthing installation. That's sort of my point here.

Ok, so... User with repository should be fine. For that user with deb file, just add [WebUpd8 repository](https://launchpad.net/~nilarimogard/+archive/ubuntu/webupd8/) and let it autoupdate in future. For Windows... oh gods :D

I mean, considering state of things on Windows, I may as well add silent autoupdate. But we have word for programs that do that im my native language, and it's not nice word.
How about adding "Download" button into that [message we talked about?](https://cloud.githubusercontent.com/assets/1282500/17014056/a502b684-4f22-11e6-9aab-6e5dd7aad4cc.png)

**Comment by buttle** [comment]
> How about adding "Download" button into that message we talked about?

That is definitely a step forward :)

At the moment, I think offering IT support to a geographically disperse user group is not practical. However, that magical 'force to latest' option would make it possible because at least they would be syncing correctly.

I would like to setup ST for users, and I feel that some way to ensure that everyone is compatible (on the same syncthing version) is really important because 1) syncthing works and 2) IT support becomes easier.

I'm sure you get the idea and it's not about me go on and on about it here. I imagine there are difficulties but please take into consideration my thoughts.
Most users are not techies. The more human intervention required, the harder it will be to make syncthing go mainstream.

If I can be of any help...
:)
Chris.
By the way, I really like what you've done to date.

**Comment by kozec** [comment]
> However, that magical 'force to latest' option would make it possible because at least they would be syncing correctly.

Currently, if you have daemon updates enabled, ST-GTK *will* force Syncthing to latest version, but only twice per day. In fact, you can force update check by opening UI Settings and clicking OK, but that's undocumented "feature" :)

What you are proposing is, let's say,  `Check for Syncthing Updates` option in menu?

**Comment by buttle** [comment]
> What you are proposing is, let's say, Check for Syncthing Updates option in menu?

Yes, but not only minor updates, major versions too.

The ST api doesn't seem to change that much so, although I know you feel that can break ST-GTK, I think the benefits outweigh the possible problems. We need to keep everybody up to date.

If a user updates various ST major versions, eventually that will cause problems with  ST-GTK. When that happens the users who do not use a package manager will have to update ST-GTK manually, but at least in the meantime, their folders are synced with everyone else.

How do you feel about that?

**Comment by kozec** [comment]
Wait, wait, I don't "feel" that API change may break something, I know it. I'm testing each version before allowing ST-GTK to download it :D

Allowing user to download non-compatible version of Syncthing through UI is definitely bad idea and your average non-technical user may easily end in state when nothing works and he don't have way how to fix it.

So, what's possible is that `Check for Syncthing Updates` button, that would just speed up autoupdate process (and display some message if there is nothing to update), but in case when ST-GTK update is needed, nothing else but ST-GTK update can help.

**Comment by buttle** [comment]
> but in case when ST-GTK update is needed, nothing else but ST-GTK update can help.

Ok, I understand completely.

My problem is how to keep a group updating to the same version. It seems like webUI only, might be the only solution... :(

Cheers.

**Comment by kozec** [comment]
> My problem is how to keep a group updating to the same version. 

But if you are able to keep them at same ST-GTK version, they'll end with same, last supported Syncthing version as well.

**Comment by buttle** [comment]
> But if you are able to keep them at same ST-GTK version

Unfortunately I can't. And even if I could, I have no control over the friends of the friends.

ST version incompatibility is the problem and the best way to solve that is to have everybody on the same version. Which version? The latest version.

ST-GTK relies on 1) a package manager or 2) tech savy users. ST compatibility can't be guaranteed. :(

I don't know, any ideas?
Chris.

**Comment by CallumCameron** [comment]
I think [this message](https://cloud.githubusercontent.com/assets/1282500/17014056/a502b684-4f22-11e6-9aab-6e5dd7aad4cc.png) is a good idea, especially on Windows (and any other platform without a package manager). It might also be useful to display a notification on the desktop so that users know they need to upgrade ST-GTK, even if they never open the ST-GTK window. The update check could be disabled in the versions compiled for package managers, like syncthing does with the web UI's update option, to avoid confusion.

Keeping a cluster with different platforms up to date is a bit awkward at the moment. I have Ubuntu and Windows machines on mine; the Ubuntu machines all upgrade through the package manager, and unless I realise there has been an incompatible protocol change, it can be weeks before anyone notices the Windows machines are no longer syncing. At least with a notification, the Windows users would know they need to upgrade ST-GTK, even if they still have to do it manually.

Doesn't help with @buttle's issue of *always* keeping everyone on the latest ST version, but at least it would help everyone get to the latest version *eventually*.

**Comment by fbatogo** [comment]
If I can chime in here with my own experience.

I have a dozen or two devices syncing different things with syncthing.   Up until a few weeks ago, keeping them up to date was pretty easy, as I had access to all of them.  (All but one of them were machines in my house.  The one outside my house was configured with OpenVPN to connect back to my house so I could change it as needed.)

However, I recently added a sync to my sister's computer.   Then, a new major version of Syncthing dropped.   I tried to keep all my devices on the older version until I could update her machine, but I have a few Android devices syncing, and they get updated automatically.  (Yeah, I could disable that, but I got the update before I thought about it.)

My sister is moderately tech savvy, so I e-mailed her and asked her to restart SyncthingGTK, thinking it would update.   But, after reading the messages here, it sounds like it won't.

I agree with whatever the bad word is for programs silently updating themselves.   Few things are worse than a silent update that breaks something as you won't know that it was an update that broke it!  But, making it a bit easier to upgrade SyncthingGTK would help a bunch!

I think showing the message that a new version is available is a good start.  If you could click a button to start an in-place upgrade, that would be perfect!   However, even being able to click a button to download the latest version would be a useful solution.   (Or, even better click the button to download the new version, and then prompt the user if they want to install it now.   That would be the best of both worlds!)

At the same time, for someone like my sister that rarely opens the UI or reboots her computer, it would be nice to have an option for a notification "bubble" to come up from time-to-time telling her that she should upgrade SyncthingGTK.   I generally hate those bubbles, but I think if it was an opt-in setting, and there was a way to tell the bubble to go away forever, it would be an acceptable solution.

Also, thanks for the app!   I used to use BTSync, and knew that without a tray icon or app my wife would never get on-board with Syncthing.   SyncthingGTK gave her what she needed to feel comfortable.  I then loaded it on the boxes I use that aren't headless because I like the visual feedback from the tray icon.

**Comment by dr4Ke** [comment]
I'm using syncthing-gtk on windows since syncthing 0.12. I upgraded to syncthing 0.13 through version 0.9.0.1. It worked fine.
For syncthing 0.14, I thought I would do the same: upgrade syncthing-gtk to 0.9.2 and it would upgrade syncthing to 0.14. But it didn't.
Is there a simple way to upgrade syncthing from 0.13 to 0.14 from syncthing-gtk?

(I know I can just replace the syncthing executable, but it will be very difficult to reproduce for my association coworkers that don't have any experience in computer administration)

**Comment by kozec** [comment]
@dr4Ke do you have "Keep syncthing up-to-date" option enabled in UI Settings? It's on Daemon page.

If yes, 0.9.2 _is_ compatible down to 0.13, so it's maybe just waiting for next updatecheck - it occurs only once per day.

**Comment by dr4Ke** [comment]
OK. It must be that! I had not used this computer for a few weeks, and it did a syncthing upgrade (0.13.x) just before I installed the latest syncthing-gtk. I'll check that tonight.
